fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreRepairHeaderTest.java

```diff
@@ -0,0 +1,642 @@
+        FDBRecordStoreBase.UserVersionChecker userVersionChecker = new AssertMatchingUserVersion(0, 1);
+
+        try (FDBRecordContext context = openContext()) {
+            getStoreBuilder(context, recordMetaData)
+                    .setUserVersionChecker(userVersionChecker)
+                    .open()
+                    .setStateCacheability(true);
+            commit(context);
+        }
+
+        userVersionChecker = new AssertMatchingUserVersion(1, 1);
+
+        timer.reset();
+        try (FDBRecordContext context = openContext()) {
+            getStoreBuilder(context, recordMetaData)
+                    .setUserVersionChecker(userVersionChecker)
+                    .open();
+            assertEquals(1, timer.getCount(FDBStoreTimer.Counts.STORE_STATE_CACHE_MISS));
+            commit(context);
+        }
+
+        timer.reset();
+        try (FDBRecordContext context = openContext()) {
+            getStoreBuilder(context, recordMetaData)
+                    .setUserVersionChecker(userVersionChecker)
+                    .open();
+            assertEquals(1, timer.getCount(FDBStoreTimer.Counts.STORE_STATE_CACHE_HIT));
+            commit(context);
+        }
+
+        timer.reset();
+        return userVersionChecker;
+    }
+
+    private List<Tuple> createInitialStore(final FormatVersion initialFormatVersion, final RecordMetaDataProvider metaData) {
+        try (FDBRecordContext context = openContext()) {
+            recordStore = getStoreBuilder(context, metaData, path)
+                    .setFormatVersion(initialFormatVersion)
+                    .createOrOpen();
+            TestRecords1Proto.MySimpleRecord rec = TestRecords1Proto.MySimpleRecord.newBuilder()
+                    .setRecNo(1)
+                    .setStrValueIndexed("a_1")
+                    .setNumValueUnique(1)
+                    .setNumValue2(7)
+                    .build();
+            final FDBStoredRecord<Message> storedRecord = recordStore.saveRecord(rec);
+            commit(context);
+            return List.of(storedRecord.getPrimaryKey());
+        }
+    }
+
+    private static class AssertMatchingUserVersion implements FDBRecordStoreBase.UserVersionChecker {
+        private final int oldUserVersion;
+        private final int newUserVersion;
+
+        public AssertMatchingUserVersion(final int oldUserVersion, final int newUserVersion) {
+            this.oldUserVersion = oldUserVersion;
+            this.newUserVersion = newUserVersion;
+        }
+
+        @Override
+        public CompletableFuture<Integer> checkUserVersion(@Nonnull final RecordMetaDataProto.DataStoreInfo storeHeader, final RecordMetaDataProvider metaData) {
+            assertEquals(oldUserVersion, storeHeader.getUserVersion());
+            return CompletableFuture.completedFuture(newUserVersion);
+        }
+
+        @Override
+        @SuppressWarnings("deprecation") // overriding deprecated method
```

Comment:
If you wanted, you could probably implement these deprecated variants with `return fail()`, as we should never call them within check version